pubsub/envelope.go

```diff
@@ -110,6 +116,29 @@ func NewCloudEventsEnvelope(id, source, eventType, subject string, topic string,
 	}
 }
 
+// HasExpired determines if the current cloud event has expired.
+func (cloudEvent *CloudEventsEnvelope) HasExpired() bool {
+	if cloudEvent.Expiration != "" {
+		expiration, err := time.Parse(time.RFC3339, cloudEvent.Expiration)
+		if err != nil {
+			return false
+		}
+
+		return expiration.UTC().Before(time.Now().UTC())
+	}
+
+	return false
+}
+
+// ApplyMetadata will process metadata to modify the cloud event accordingly
+func (cloudEvent *CloudEventsEnvelope) ApplyMetadata(metadata map[string]string) {
+	ttl, hasTTL, _ := contrib_metadata.TryGetTTL(metadata)
+	if hasTTL {
+		// Adds an extra second to allow components to expire message before Dapr.
+		cloudEvent.Expiration = time.Now().UTC().Add(ttl).Add(time.Second).Format(time.RFC3339)
```

Comment:
Potential corner-case: when the time is set to the maximum possible time, adding one second would overflow to the past. Programmers tend to use constants like that to indicate that a message should never expire.

Reply:
Good point.

Reply:
I actually found an overflow issue in the logic to parse Duration from seconds. It is virtually impossible for an overflow to happen here because the max duration is `maxInt64`, which is not enough to overflow `Now().Add(ttl)` for looong time.